File: combine-handler/src/main/java/cc/vileda/kalfor/handler/SchemaValidationHandler.java
package cc.vileda.kalfor.handler;

import io.vertx.core.Handler;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import io.vertx.core.logging.Logger;
import io.vertx.core.logging.LoggerFactory;
import io.vertx.rxjava.ext.web.RoutingContext;
import org.everit.json.schema.Schema;
import org.everit.json.schema.ValidationException;
import org.everit.json.schema.loader.SchemaLoader;
import org.json.JSONArray;
import org.json.JSONObject;
import org.json.JSONTokener;

import java.io.IOException;
import java.io.InputStream;


public class SchemaValidationHandler implements Handler<RoutingContext>
{
	private static final Logger LOGGER = LoggerFactory.getLogger(SchemaValidationHandler.class);
	private final Schema schema;

	public SchemaValidationHandler()
	{
		try (InputStream inputStream = getClass().getResourceAsStream("/kalfor-schema.json")) {
			final JSONObject rawSchema = new JSONObject(new JSONTokener(inputStream));
			schema = SchemaLoader.load(rawSchema);
		}
		catch (IOException e) {
			LOGGER.error(e.getMessage(), e);
			throw new RuntimeException(e);
		}
	}

	@Override
	public void handle(final RoutingContext routingContext)
	{
		try {
			schema.validate(new JSONArray(routingContext.getBodyAsString()));
			routingContext.next();
		} catch (final ValidationException validationException) {
			final JsonObject error = new JsonObject().put("error", validationException.getMessage());
			final JsonArray errorMessages = new JsonArray();

			validationException.getCausingExceptions()
					.forEach(e -> errorMessages.add(e.getPointerToViolation()));

			error.put("problems", errorMessages);
			routingContext.response()
					.setStatusCode(405)
					.end(error.encodePrettily());
		}
	}
}

File: handler/src/main/kotlin/cc/vileda/kalfor/handler/SchemaValidationHandler.kt
package cc.vileda.kalfor.handler

import com.github.fge.jackson.JsonLoader
import com.github.fge.jsonschema.core.exceptions.ProcessingException
import com.github.fge.jsonschema.core.report.ProcessingReport
import com.github.fge.jsonschema.main.JsonSchema
import com.github.fge.jsonschema.main.JsonSchemaFactory
import io.vertx.core.Handler
import io.vertx.core.json.JsonArray
import io.vertx.core.json.JsonObject
import io.vertx.core.logging.Logger
import io.vertx.core.logging.LoggerFactory
import io.vertx.rxjava.ext.web.RoutingContext

import java.io.IOException
import java.net.HttpURLConnection


class SchemaValidationHandler : Handler<RoutingContext> {
    private val schema: JsonSchema

    init {
        try {
            schema = JsonSchemaFactory.byDefault().getJsonSchema("resource:/kalfor-schema.json")
        } catch (e: ProcessingException) {
            LOGGER.error(e)
            throw RuntimeException(e)
        }

    }

    override fun handle(routingContext: RoutingContext) {
        try {
            val validate = schema.validate(JsonLoader.fromString(routingContext.bodyAsString))
            if (validate.isSuccess) {
                routingContext.next()
                return
            }

            val messages = JsonArray()
            validate.forEach { processingMessage -> messages.add(processingMessage.message) }
            val error = JsonObject().put("error", messages)
            routingContext.response().setStatusCode(HttpURLConnection.HTTP_BAD_REQUEST).end(error.encodePrettily())
        } catch (e: ProcessingException) {
            LOGGER.error(e)
            val error = JsonObject().put("error", e.message)
            routingContext.response().setStatusCode(HttpURLConnection.HTTP_BAD_REQUEST).end(error.encodePrettily())
        } catch (e: IOException) {
            LOGGER.error(e)
            val error = JsonObject().put("error", e.message)
            routingContext.response().setStatusCode(HttpURLConnection.HTTP_BAD_REQUEST).end(error.encodePrettily())
        }

    }

    companion object {
        private val LOGGER = LoggerFactory.getLogger(SchemaValidationHandler::class.java)
    }
}
